### ISSUE TITLE
Window resize drag location is inside the window

### ISSUE THREAD

**Issue by shujaatak** [bitsdojo/bitsdojo_window#110]
Tested in Ubuntu 20.04.3 LTS.

The borders used for resizing the window are inside the window so any widget which is near the edges of the window can't interact with mouse events.

The window should be re-sizable only when the mouse is near but outside of the visible window.

**Comment by hBurt** [comment]
So, @j787701730  is actually moving in the right direction here. @shujaatak the fix is below

tl;dr : scroll down for the fix, window is big and we make window smaller


# The Explination

The red border on this image are the areas where the resizing cursor is displayed (Currently inside the application)
![image](https://user-images.githubusercontent.com/4024074/140266574-1999a6e0-dcf2-43d3-b09b-a93fc232bff7.png)

Since that is currently inside the application, we need a way to move it out. Well, you cant really extend the window out past its original size without just making the window larger itself, and by doing that you continue to resize WM_NCHITTEST as well. 

The way to fix this is by editing NCCALCSIZE_PARAMS from lparam.

What this essentially does is make the acutal window smaller, and creates an invisible area around the window which is used to scale the application. It's similar to WS_THICKFRAME from: [https://docs.microsoft.com/en-us/windows/win32/winmsg/window-styles](https://docs.microsoft.com/en-us/windows/win32/winmsg/window-styles)

# The Fix

Inside of bitsdojo_window.cpp, similar to how @shujaatak was explaining, you should edit it as such(notice the + and - 9)
```
        if (pt.x < resizeMargin - 9)
        {
            return HTLEFT;
        }
        if (pt.x > (rc.right - resizeMargin + 9))
        {
            return HTRIGHT;
        }
```

also, inside the handle_nccalcsize method add the following code underneath `params->rgrc[0].top -= 1;` which is in the methods last else statement 
```
params->rgrc[0].left += 8;
params->rgrc[0].right -= 8;
```

Adding this in resolves the issue mentioned:
![border](https://user-images.githubusercontent.com/4024074/140268747-e41587dc-f3a6-4ec5-867d-792f1700584e.gif)

# Small issues

This will not fix resizing for the top/ bottom/corners since I didn't change the `HTBOTTOM` and `HTTOP`. I am however confident that with the information I've provided you will be able to figure out the rest. You will also have to modify `params->rgrc[0].bottom -= 8;` to get the bottom to work. 

I would like to mention that if you modify `params->rgrc[0].top -= 1;` to be larger, you will start to see the origional windows: minimize, maximize, and close buttons. I know that `WS_POPUP` removes the frame but haven't messed around with it. You would add styles like that to this line: `style = style | WS_CLIPCHILDREN | WS_POPUP;`

For me, there is a 1px padding on the right hand side of the window between the window buttons and the edge of the window(Not seen in the above gif because it goes away when the window isn't active) . This is due to incorrect sizing (the hardcoded ints: 8 and 9). What we should really be doing here is scaling the NCCALCSIZE_PARAMS off of the DPI.


## Good luck!

**Comment by shujaatak** [comment]
@hBurt What an awesome comment! I would definitely try your suggestions soon!

**Comment by shujaatak** [comment]
@hBurt I wonder if your solution produces the window frame shadow too?

**Comment by PeterNjeim** [comment]
My solution:
```diff
# bitsdojo_window/bitsdojo_window_windows/windows/bitsdojo_window.cpp
@@ -150,8 +150,14 @@
         DwmExtendFrameIntoClientArea(hwnd, &margins);
     }
 
+    double getScaleFactor(HWND window) {
+        UINT dpi = GetDpiForWindow(window);
+        return dpi / 96.0;
+    }
+
     LRESULT handle_nchittest(HWND window, WPARAM wparam, LPARAM lparam)
     {
+        double scaleFactor = getScaleFactor(window);
         bool isMaximized = IsZoomed(flutter_window);
         if (isMaximized)
             return HTCLIENT;
@@ -162,33 +168,33 @@
         int resizeMargin = getResizeMargin(window);
         if (pt.y < resizeMargin)
         {
-            if (pt.x < resizeMargin)
+            if (pt.x < resizeMargin - (int)ceil(scaleFactor * 5))
             {
                 return HTTOPLEFT;
             }
-            if (pt.x > (rc.right - resizeMargin))
+            if (pt.x > (rc.right - resizeMargin + (int)ceil(scaleFactor * 5)))
             {
                 return HTTOPRIGHT;
             }
             return HTTOP;
         }
-        if (pt.y > (rc.bottom - resizeMargin))
+        if (pt.y > (rc.bottom - resizeMargin + (int)ceil(scaleFactor * 7)))
         {
-            if (pt.x < resizeMargin)
+            if (pt.x < resizeMargin - (int)ceil(scaleFactor * 5))
             {
                 return HTBOTTOMLEFT;
             }
-            if (pt.x > (rc.right - resizeMargin))
+            if (pt.x > (rc.right - resizeMargin + (int)ceil(scaleFactor * 5)))
             {
                 return HTBOTTOMRIGHT;
             }
             return HTBOTTOM;
         }
-        if (pt.x < resizeMargin)
+        if (pt.x < resizeMargin - (int)ceil(scaleFactor * 5))
         {
             return HTLEFT;
         }
-        if (pt.x > (rc.right - resizeMargin))
+        if (pt.x > (rc.right - resizeMargin + (int)ceil(scaleFactor * 5)))
         {
             return HTRIGHT;
         }
@@ -250,11 +256,6 @@
         }        
     }
 
-    double getScaleFactor(HWND window) {
-        UINT dpi = GetDpiForWindow(window);
-        return dpi / 96.0;
-    }
-
     LRESULT handle_nccalcsize(HWND window, WPARAM wparam, LPARAM lparam)
     {
         auto params = reinterpret_cast<NCCALCSIZE_PARAMS*>(lparam);
@@ -286,7 +287,10 @@
         }
         else 
         {
-            params->rgrc[0].top -= 1;
+            params->rgrc[0].top -= (int)ceil(scaleFactor * 1);
+            params->rgrc[0].bottom -= (int)ceil(scaleFactor * 6);
+            params->rgrc[0].left += (int)ceil(scaleFactor * 4);
+            params->rgrc[0].right -= (int)ceil(scaleFactor * 4);
         }
 
         return 0;
```

**Comment by PeterNjeim** [comment]
> @PeterNjeim, in your implementation there was still a white border around the app which looked bad in dark-mode.
> 
> I have raised #139, looking for any review.

I have no white border in my app.

**Comment by alexmercerind** [comment]
> > @PeterNjeim, in your implementation there was still a white border around the app which looked bad in dark-mode.
> > I have raised #139, looking for any review.
> 
> I have no white border in my app.

Possibly you're just testing Windows 11.

**Comment by hruzgar** [comment]
I just found this thread after encountering the same problem (resize drag being inside the window). Does any of the fixes actually work? @PeterNjeim @alexmercerind

**Comment by alexmercerind** [comment]
> I just found this thread after encountering the same problem (resize drag being inside the window).
> Does any of the fixes actually work? [@PeterNjeim](https://github.com/PeterNjeim) [@alexmercerind](https://github.com/alexmercerind)

At the end I created my own plugin, that I use in my project: https://github.com/alexmercerind/window_plus

Doesn't have everything, but is something I can trust.

You may also consider using [package:window_manager](https://pub.dev/packages/window_manager), which is another package.

**Comment by hruzgar** [comment]
It's nice to see that I'm not the only one having these problems. I will definitely check out these packages thank you so much!
 
Btw do you know if `window_manager` handles the issue with the resize drag being inside the window by default or does it also need some custom logic?

The only things I need for my application for now are custom window controls (which was currently possible with `bitsdojo_window`) and maybe also change the Fullscreen/Size Down Icon with a callback function or some other logic. Other than this specific issue with "drag-to-resize", I don't really have the need for other functionality. Given my relatively simple needs, could you give me a recommendation on which of these two packages to choose?

**Comment by hruzgar** [comment]
So for anyone wondering what I did to overcome this issue: 
- previosly I was using bitsdojo_window. Everything was working ok (not perfect), but I realized that the resize drag icon was inside the window and there is no easy fix.
- now i switched to [window_manager](https://pub.dev/packages/window_manager) and it's just perfect. 
	1. It has the resize drag outside the window
	2. All the other stuff that I had with bitsdojo_window also works fine (I just had to change `onPressed` for my custom control buttons)
	3. And I could even improve the icon change handling because with `window_manager` you can override the window event methods like `onWindowMaximize` and `onWindowUnmaximize`

Overall as a summary, everything is working perfectly now with `window_manager` and I'm happy what I ended up with. (on Windows at least, idk what will happen with macos and Linux..)

Thank you again @alexmercerind for the recommendation
